Clojure Contrib Package

gigasquid opened the pull request:
## Description ##
This adds a Clojure package for MXNet under a contrib package

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ X] Changes are complete (i.e. I finished coding on this PR)
- [ X] All changes have test coverage:
- [ X] Code is well-documented: 
- [ X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
This is the initial commit of the Clojure package in as a contrib project. There will be a period of stabilization, feedback, and contribution. After that, it may graduate to out of contrib.

The initial package has been tested by the community in a separate repo prior to this PR https://github.com/gigasquid/clojure-mxnet

Here is the original issue:
https://github.com/apache/incubator-mxnet/issues/8971

Architecture & Design:
https://cwiki.apache.org/confluence/display/MXNET/MXNet+Clojure

Current List of Package Completion and Contribution Needs
https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs

## Comments ##
There have been areas of improvement noted from feedback on the mailing list.

- Increased test coverage
- Benchmarks

In addition there will be a need to integrate with the Makefile and CI. 

I would like to propose that this initial PR be merged as it represents a working, community tested development milestone and then add on the improvements as separate PRs to allow smaller commits, better visibility, and community involvement.

gigasquid commented:
@nswamy This is ready for review/feedback

lanking520 commented:
Hi Carin, is there anyway I can test it? I am pretty new to Clojure

kurman commented:
Do you think you can split examples from cde0ee3a72426946e86b6ba0154d813a1a5611e3 into a separate commit to make the review process easier?

gigasquid commented:
@kurman sure - I will rewrite the commits and and push tomorrow (wrapping up today 😄 )

gigasquid commented:
@kurman - I rewrote the commits into 2 parts - one with the main package and tests and another with the examples. Please let me know if I can do anything else to make the review process easier.

kurman commented:
@gigasquid thanks for splitting the commit, I am already finding that it is easer going through the PR.

kovasb commented:
Thanks for this contribution!

In terms of code quality this looks very good, getting a lot done in a straightforward manner. People should feel good shipping this if/when test coverage is satisfactory.

My only code question is: Where do the corner cases tend to lie? Whats the most complicated part where bugs might be lurking?

Some higher level questions:

- What is the debug experience like? It seems like errors can come from 3 places (mxnet core, scala, or clojure), is this a problem in practice? Can users track down what they did wrong?

- What does maintenance look like? As MXNet & Scala API evolve, how much work does this codebase need to do to keep up? What changes are picked up automatically via introspection, vs requiring changing the codebase?

- How do people learn to use the thing? Is it easy to go between the clojure functions and existing MXNet docs/tutorials? How is the documentation going to work long term?

gigasquid commented:
@kovasb Thanks so much for taking the time to review and providing such insightful feedback and questions.

As far as corner cases go, I feel pretty confident about the interop with the Scala api. I would say the risk for complications would be lurking in the areas that I haven't fully ported over yet and explored. I ran into some trouble with the protocols for the custom operators with class loading and deferred working on it more to focus on some other areas. I'm hoping it was just some silly thing that I wasn't doing properly because in theory, if it's a Scala/Java class - it can all be interoped with, but I won't be 100% sure until I dive in again.

You have some excellent higher level questions:

* The debug experience is an area that could use improvement. I've added it as an item on the https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs page. The regular JVM stack traces that occur in clojure point to the errors in the Scala or Clojure class and can be pinned down pretty readily and are not bad. The more frustrating ones that I have encountered is when there is a core dump, (for example binding some layers with a bad shape), in the mxnet-core. You can still find the line where the error occurred in the underlying Scala/Clojure code, but it is not easy to navigate. From what I understand, I believe that this might be the same case in the current Scala package as well, so I'm thinking that working together we could make it better for both packages if we could tackle this.

* I think the story for maintenance is pretty good because of leveraging the Scala package. The Clojure package also has the flexibility to evolve at its own rate by deciding when to move to the next Scala version or wrap a new Scala api.  I think the philosophy in MXNet is to not make breaking changes is very good and will enable stability. The mxnet core api changes for NDArray and Symbol flow through the Scala package via the c code generators. The clojure package picks these changes up from the Scala classes by running its own generator, (which is a simple lein command). Changes to the higher level Scala apis will also be picked up or course, while any new apis would need to be added and incorporated manually.

* The Clojure apis have been developed with the goal of being intuitive from the existing apis while also being idiomatic to use from the Clojure perspective. I've been able to use existing tutorials from both Python and Scala and translate them to Clojure. Here is an example:

From the Python tutorial https://mxnet.incubator.apache.org/tutorials/r/symbol.html you have:

```python
net <- mx.symbol.Variable("data")
net <- mx.symbol.FullyConnected(data=net, name="fc1", num_hidden=128)
net <- mx.symbol.Activation(data=net, name="relu1", act_type="relu")
net <- mx.symbol.FullyConnected(data=net, name="fc2", num_hidden=64)
net <- mx.symbol.Softmax(data=net, name="out”)
```


Translation to clojure follows pretty directly from this:

```clojure
(def data (sym/variable "data"))
(def fc1 (sym/fully-connected "fc1" {:data data :num-hidden 128}))
(def act1 (sym/activation "act1" {:data fc1 :act-type "relu"}))
(def fc2 (sym/fully-connected "fc2" {:data act1 :num-hidden 64}))
(def net (sym/softmax-output "out" {:data fc2}))
```

But also can be written in a more idiomatic, threading way using the `as->` clojure function

```clojure
(as-> (sym/variable "data") data
  (sym/fully-connected "fc1" {:data data :num-hidden 128})
  (sym/activation "act1" {:data data :act-type "relu"})
  (sym/fully-connected "fc2" {:data data :num-hidden 64})
  (sym/softmax-output "out" {:data data}))
```
As far as documentation goes, the design choice has been made to generate the code from the symbol and and ndarray apis and persist that to files rather than read into memory as macros so that it enable automated API documentation with Codox.  Going forward, I envision a section of the MXNet webpage with tutorials and API documentation similar to the other language bindings to also help the user getting started.

larroy commented on contrib/clojure-package/examples/cnn-text-classification/get_data.sh:
```diff
@@ -0,0 +1,29 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
```
Could you add "set -e" to this script?

cjolivier01 replied:
isn’t that just below set -evx ?

larroy commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,197 @@
+### Use Prebuilt Jars
+There are deployed jars on Clojars for each supported system
+
+* `[org.apache.clojure-mxnet/clojure-mxnet-linux-gpu "0.1.1-SNAPSHOT"]`
```
Who is building these packages?   How is the version number related to the mxnet version?

Can we add instructions to build the package as well? Something like when you build a scala package in the local ivy repo.

gigasquid replied:
That's a good point. The section was taken from the README in the external git repo that I had for testing. I had built and deployed those jars manually and pushed them to clojars.

I'm not sure how the build will happen within the project when it gets merged long term.  We will need some discussion on that. For the short term, I can continue to build the jars manually and push them to clojars for general users to consume and give feedback on. 

I definitely can document the process that I use to do the build and deploy 👍

larroy commented on contrib/clojure-package/examples/pre-trained-models/download-resnet-152.sh:
```diff
@@ -0,0 +1,27 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
```
set -e

larroy commented on contrib/clojure-package/examples/pre-trained-models/download-caltech.sh:
```diff
@@ -0,0 +1,25 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
```
set -e

gigasquid commented:
Thanks for taking the time to review @larroy 😸 I'll get to work on incorporating your feedback.

larroy commented:
Hi again @gigasquid thanks a lot for your contribution. Some of us are not totally familiar with Clojure, so we are trying to find someone to help us review it. I have to say your PR has gotten a lot of people excited and we are very enthusiastic about this contribution. I will try to find another pair of eyes with more clojure experience and get back to you. I have retriggered CI to have a clean run. Do you think it would make sense to add a clojure stage to our CI pipeline similar to the Scala one for these bindings to test them?

gigasquid commented:
@larroy and @yzhliu  - Thanks. Please let me know if there is anything I can do to help the review and get this merged. I've been trying to keep this PR in sync in the external github repo with other's contributions and generally improving the test coverage, but I can stop adding commits and freeze it to give you a stable PR to review and merge, if that helps. Just let me know.

To recap the work and status of the PR so far:

- It has been reviewed by @kovasb - who is experienced in both Clojure and Deep Learning
- The feedback from your review, @larroy, has been implemented
- A code coverage plugin has been integrated
- A few bug fix contributions have been brought over from the main repo
- The scala jar version has been bumped to `1.2.1` (from using the staging jars)
- Additional unit tests have been added to increase coverage. The current stats are:

```
|-----------------------------------------+---------+---------|
|                               Namespace | % Forms | % Lines |
|-----------------------------------------+---------+---------|
|                           dev.generator |   93.07 |   94.33 |
|           org.apache.clojure-mxnet.base |  100.00 |  100.00 |
|       org.apache.clojure-mxnet.callback |  100.00 |  100.00 |
|        org.apache.clojure-mxnet.context |   76.00 |   85.71 |
|          org.apache.clojure-mxnet.dtype |  100.00 |  100.00 |
|    org.apache.clojure-mxnet.eval-metric |   95.00 |  100.00 |
|       org.apache.clojure-mxnet.executor |   89.66 |   95.00 |
|    org.apache.clojure-mxnet.initializer |   25.00 |   43.75 |
|             org.apache.clojure-mxnet.io |   71.71 |   80.00 |
|        org.apache.clojure-mxnet.kvstore |   82.91 |   76.00 |
| org.apache.clojure-mxnet.kvstore-server |   58.73 |   50.00 |
|   org.apache.clojure-mxnet.lr-scheduler |   22.22 |   66.67 |
|         org.apache.clojure-mxnet.module |   74.99 |   88.62 |
|        org.apache.clojure-mxnet.monitor |  100.00 |  100.00 |
|        org.apache.clojure-mxnet.ndarray |   90.86 |   91.38 |
|      org.apache.clojure-mxnet.optimizer |   91.21 |   88.24 |
|       org.apache.clojure-mxnet.profiler |   20.00 |   40.00 |
|         org.apache.clojure-mxnet.random |  100.00 |  100.00 |
|          org.apache.clojure-mxnet.shape |   94.44 |   90.00 |
|         org.apache.clojure-mxnet.symbol |   48.87 |   54.62 |
|           org.apache.clojure-mxnet.util |   98.86 |   99.16 |
|  org.apache.clojure-mxnet.visualization |    6.82 |   20.00 |
|-----------------------------------------+---------+---------|
|                               ALL FILES |   80.01 |   84.14 |
|-----------------------------------------+---------+---------|
```

I feel the tests are pretty stable, so if you want to add a clojure stage to the CI pipeline, it would be great. All that should be needed is to be able to run the build/test tool `lein` (https://leiningen.org/).
To be specific:

`cd contrib/clojure-package`
`lein test`

It would be great to get this merged so that other people can start getting involved and helping to grow and improve the package. 😸

nswamy commented:
@gigasquid I plan to spend sometime over the weekend to go through the code.

gigasquid commented:
@larroy I just looked at the Jenkins CI pipeline. I see there is a scala cpu and gpu tests right now. For the current clojure package, the tests are all cpu. I'm not sure what system the CI is running on (OSX or Linux)? If it's linux then the Scala jar dependency will need to be changed in the `project.clj` file. Currently, it is using OSX 

```clojure
 ;; Choose the right dependency for your system
                 [org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
                 ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.2.1"]
                 ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.2.1"]
```
To choose the linux cpu, you would change this to

```clojure
 ;; Choose the right dependency for your system
                 ;[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
                 [org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.2.1"]
                 ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.2.1"]
                
```

nswamy commented:
Its running on Ubuntu Linux inside a docker container

On Fri, Jun 22, 2018 at 2:35 PM, Carin Meier <notifications@github.com>
wrote:

> @larroy <https://github.com/larroy> I just looked at the Jenkins CI
> pipeline. I see there is a scala cpu and gpu tests right now. For the
> current clojure package, the tests are all cpu. I'm not sure what system
> the CI is running on (OSX or Linux)? If it's linux then the Scala jar
> dependency will need to be changed in the project.clj file. Currently, it
> is using OSX
>
>  ;; Choose the right dependency for your system
>                  [org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
>                  ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.2.1"]
>                  ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.2.1"]
>
> To choose the linux cpu, you would change this to
>
>  ;; Choose the right dependency for your system
>                  ;[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
>                  [org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.2.1"]
>                  ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.2.1"]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/pull/11205#issuecomment-399588943>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABJxQsydxjj1SLUWB6jBJ9Ioyn0Ej7wcks5t_WMagaJpZM4Ugvws>
> .
>

didibus commented:
Shouldn't the project versions no longer be SNAPSHOT if they are to be pulled in the incubator? My understanding is that Maven will look for the newest release artifact, so different pull of the dependencies of x.x.x.SNAPSHOT result in you possibly getting different artifacts.

But at this stage, I'd expect the APIs to be a bit more stable that changes to them warrant some version change. 0.1.1 could become 0.1.2, etc. Or some scheme with alpha: 0.1.1-alpha1 could become 0.1.1-alpha2 etc.

gigasquid commented:
@nswamy @larroy I switched the project.clj to use the linux cpu version to make it ready for CI.

@didibus - good point. A discussion needs to made concerning the versioning of the clojure-package. Right now the dependencies and main project version is in the `project.clj` file 

```clojure
(defproject org.apache.mxnet.contrib.clojure/clojure-mxnet "1.2.1-SNAPSHOT"
  :description "Clojure package for MXNet"
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [t6/from-scala "0.3.0"]
                 ;; Choose the right dependency for your system
                 ;[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
                 [org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.2.1"]
                 ;[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.2.1"]
                 [org.clojure/tools.logging "0.4.0"]
                 [org.apache.logging.log4j/log4j-core "2.8.1"]
                 [org.apache.logging.log4j/log4j-api "2.8.1"]
                 [org.slf4j/slf4j-log4j12 "1.7.25" :exclusions [org.slf4j/slf4j-api]]]
```
The dependencies used are the Scala `1.2.1` jars. If the clojure project would be deployed to clojars or nexus, the `org.apache.mxnet.contrib.clojure/clojure-mxnet "1.2.1-SNAPSHOT"` would have to be changed to `org.apache.mxnet.contrib.clojure/clojure-mxnet "1.2.1"` first and then would have to be incremented.

For a point of reference, the Scala package is now sitting at `1.3.0-SNAPSHOT`. So maybe the clojure package should be set to that as well.

We can choose to have the clojure-package align with the main package version or have its own version. I'm fine with whatever we come up with. 

With a new language package there will be some process stuff to work through on how to get into CI and handle versioning deployment. I expect and encourage some conversations to be had about how best to handle it.

gigasquid commented:
@didibus I think I missed a point you were making in your last comment about snapshot versions in regards to the README documentation of the test versions I had out on clojars. That's a good point. I updated the README in this commit https://github.com/apache/incubator-mxnet/pull/11205/commits/3c48fa3cd505e0a7a2b0eb857975085e52669453 and removed the reference. I won't redeploy anywhere until we've come up with the correct versioning/ deployment/ timing process.

kurman commented on contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/classifier.clj:
```diff
@@ -0,0 +1,112 @@
+        num-label 2
+        dropout 0.5
+        input-x (sym/variable "data")
+        polled-outputs (mapv (fn [filter-size]
```
Can we extract named fn that we use in mapv?

kurman commented on contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/classifier.clj:
```diff
@@ -0,0 +1,112 @@
+        dropout 0.5
+        input-x (sym/variable "data")
+        polled-outputs (mapv (fn [filter-size]
+                               (as-> (sym/convolution {:data input-x
```
I think the alignment is messed up here, I expect:
`
(as->  exp

           name

           form1

           form2)
`

kurman commented on contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/classifier.clj:
```diff
@@ -0,0 +1,112 @@
+                                                       :num-filter num-filter}) data
+                                 (sym/activation {:data data :act-type "relu"})
+                                 (sym/pooling {:data data
+                                               :pool-type "max"
```
Is it worth having separate const namespace?

gigasquid replied:
not sure exactly what you are proposing..

kurman commented on contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/classifier.clj:
```diff
@@ -0,0 +1,112 @@
+        total-filters (* num-filter (count filter-list))
+        concat (sym/concat "concat" nil polled-outputs {:dim 1})
+        hpool (sym/reshape "hpool" {:data concat :target-shape [batch-size total-filters]})
+        hdrop (if (> dropout 0) (sym/dropout "hdrop" {:data hpool :p dropout}) hpool)
```
`(> dropout 0)` ->  `(pos? dropout)`

kurman commented on contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/classifier.clj:
```diff
@@ -0,0 +1,112 @@
+        sentence-size (:sentence-size ms-dataset)
+        shuffled (shuffle-data test-size ms-dataset)
+        train-data (mx-io/ndarray-iter [(get-in shuffled [:training :data])]
+                                       {:label[(get-in  shuffled [:training :label])]
```
we need a space between key and the value.
And there is an extra space between get-in and params

gigasquid commented:
Thanks @kurman - I appreciate you going through it, I realize that it is a _big_ PR. I'll wait until you are done to address the feedback 😄

gigasquid commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,203 @@
+
+
+### Deferred
+* Feed Forward API
```
I had asked about it earlier in the development cycle and didn't seen any of the online docs for the Scala API pointing to it. It was said that it was deprecated but still supported https://mxnet.incubator.apache.org/api/scala/index.html so I went the route of supporting Module first instead of the Feed Foward (Model) API.

gigasquid commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,203 @@
+The bulk of the ndarray and symbol apis are generated via java reflection into the Scala classes. To generate, use the `dev/generator.clj` file. These generated files are checked in as source, so the only time you would need to run them is if you are updated the clojure package with an updated scala jar and want to regenerate the code.
+
+To do this run the leiningen task
+`lein run -m dev.generator`
```
I did make an alias for it that I can change to refer in the README instead called `generate-code` to make it clearer for the readers

```clojure
  :aliases {"generate-code" ["run" "-m" "dev.generator"]}
```

gigasquid commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,203 @@
+See the Confluence page: https://cwiki.apache.org/confluence/display/MXNET/MXNet+Clojure
+
+## Building and Deploying Jars
+The process to build and deploy the jars currently is a manual process using the `lein` build tool and `Clojars`, the Clojure dependency hosting platform.
```
👍 I was documenting my current process, I definitely think it is worth a discussion of where to best host the jars. Nexus would also work fine long term.

gigasquid commented on contrib/clojure-package/examples/cnn-text-classification/README.md:
```diff
@@ -0,0 +1,38 @@
+The dataset can be obtained here: [https://github.com/yoonkim/CNN_sentence](https://github.com/yoonkim/CNN_sentence). The two files `rt-polarity.neg`
+and `rt-polarity.pos` must be put in a directory. For example, `data/mr-data/rt-polarity.neg`.
+
+You also must download the glove word embeddings. The suggested one to use is the smaller 50 dimension one
```
It was smaller and could fit into my laptop memory :)

gigasquid replied:
Word2vec is available in the demo as well - but I haven't been able to test that yet. I can put that on the Needs Help page https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs

lanking520 replied:
Currently, I cannot reproduce the same accuracy that Python have. There may be changes on the Scala Training code.

larroy commented:
Hi @gigasquid   CI runs stuff in docker containers.

The steps to add your code to CI would be the following:

1. Add your clojure dependencies to  `ci/docker/install/ubuntu_clojure.sh`     (check how it's done with scala)

2. You need to add one function to
`ci/docker/runtime_functions.sh` that builds your package

and then add a stage into the Jenkinsfile:

See for example this PR which adds an additional Android stage:

https://github.com/apache/incubator-mxnet/pull/11382/files#diff-58231b16fdee45a03a4ee3cf94a9f2c3L486

To test locally run:

ci/build.py --platform ubuntu_cpu --shm-size 500m /work/runtime_functions.sh unittest_ubuntu_cpu_scala

but just use the function for clojure that was just created.

ci/build.py --platform ubuntu_cpu -i      will put you inside the container, so you can check what steps you need to take and add to the runtime_functions.sh script.  

Let me know if you have any issues.

gigasquid commented:
Question for @kurman @larroy @nswamy - Someone has a PR out on the main repo for improving the style with cljfmt https://github.com/gigasquid/clojure-mxnet/pull/9. Applying the plugin would cause a large commit file diff https://github.com/gigasquid/clojure-mxnet/compare/agilecreativity-improve-style-with-cljfmt?expand=1 - But would result in uniformly formatted code for the initial package commit. Would you prefer to have this apply to this PR or wait until this PR is merged and then open a follow up one just for the formatting change? Happy to go with whatever you think is the best approach.

gigasquid commented on .gitignore:
```diff
@@ -170,3 +170,5 @@ tests/mxnet_unit_tests
 cc
 cxx
+contrib/clojure-package/test/test-ndarray.clj
+contrib/clojure-package/test/test-symbol.clj
```
those should be in the clojure package git-ignore - I'll move them

gigasquid commented:
@larroy I added the CI in this commit, if you'd like to review it https://github.com/apache/incubator-mxnet/pull/11205/commits/ad9792493dd9c2a92f04955da9de38bd6e48884c

It tests ok locally with `ci/build.py --platform ubuntu_cpu --shm-size 500m /work/runtime_functions.sh unittest_ubuntu_cpu_clojure`

I did add a stage to the Jenkins file, but it doesn't seem to be showing up when it's running against the branch PR, so I'm not sure if I'm missing something here or how to test that.

larroy commented:
both work for me

larroy commented on ci/docker/install/ubuntu_clojure.sh:
```diff
@@ -0,0 +1,33 @@
+chmod 777 lein
+sudo cp lein /usr/local/bin
+sudo apt install libcurl3
+sudo add-apt-repository ppa:timsc/opencv-3.4
```
Is the opencv version that is in ubuntu too old? Is not clear why we are adding this to me.

larroy replied:
I think it's also causing some issues in CI:

http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11205/19/pipeline

gigasquid replied:
yes - that is the issue I'm looking into. It appears that the version of the opencv that was used to build the scala jars on nexus is different than the opencv required by caffe. I'm looking into changing the unit test from the clojure package to run off of the ones generated by `make scalapkg` and `make scalainstall` instead, (which is most likely a better way to go anyway).

gigasquid commented:
@larroy The CI Clojure unit tests now pass 😸 
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11205/22/pipeline

![jenkins___incubator-mxnet___pr-11205____22](https://user-images.githubusercontent.com/340299/41974206-54d2a8dc-79e5-11e8-9702-787e16ab5223.png)

gigasquid commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,204 @@
+
+## Introduction
+
+MXNet is a first class, modern deep learning library that AWS has officially picked as its chosen library. It supports multiple languages on a first class basis and is incubating as an Apache project.
```
I'm a little confused by the +1 - I'm interpreting that it's ok to leave the AWS in the text, but if I understood that wrong, please let me know :)

larroy commented on contrib/clojure-package/src/org/apache/clojure_mxnet/module.clj:
```diff
@@ -0,0 +1,691 @@
+                                 (callback/invoke cb i batch-num eval-metric))
+                               (.dispose batch)
+                          (inc batch-num)))
+       (println "Epoch " i " Train-" (eval-metric/get eval-metric))
```
isn't it a better practice to use logging? in Scala we often use logback for logging, so you can configure logging for these classes independently.

gigasquid replied:
Thanks - excellent point. I put an item to improve that in https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs

gigasquid commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,204 @@
+
+The motivation section addresses this, but the main reason is high leverage is using the great work that the Scala package has already done.
+
+**How can I tell if the gpu is being used?**
```
Thanks I updated the README

gigasquid commented:
Thanks @lanking520 for the review :) - In regards to the CNN text classification, the example at the default is running on a limited dataset (to fit in my laptop memory). There is an item on the Needs Contribution Help page about taking another look at that example with Word2Vec. I added to compare it to the python example results as well.

yzhliu commented on contrib/clojure-package/README.md:
```diff
@@ -0,0 +1,230 @@
+
+ Run `make scalapkg` then `make scalainstall`
+
+then replace the correct jar for your architecture in the project.clj, example `[ml.dmlc.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.3.0-SNAPSHOT"]`
```
org.apache instead?

gigasquid replied:
Nice catch - will update

yzhliu commented on contrib/clojure-package/src/org/apache/clojure_mxnet/gen/ndarray.clj:
```diff
@@ -0,0 +1,2312 @@
+
+
+
+
```
in my understanding, you're listing all the operators below. Can Clojure auto-generate these functions? or can we have a function to allow users call by function name? e.g., `NDArray.call("foo", arg1, arg2, ...,)`.

lanking520 replied:
Currently, Clojure generated these by the write file method. I saw @gigasquid uploads the 3000+ lines of code on Symbol and NDArray in the Gen folder.

gigasquid replied:
@yzhliu @lanking520 That is correct. We are autogenerating the code in the `gen` directories using the `/dev/generator.clj` code. This code gets invoking using a command line `lein generate-code` and produces these files by inspecting the classes created by Scala with java reflection. Where the scala package creates these classes using macros directly into memory without persisting, the clojure package chose to persist them to files to help the user discover what functions are available, generate api documentation, and to be able to see what new functions were added via a git diff from release to release.

yzhliu replied:
Can we generate the source code file during compiling, instead of checking in to the repo?

gigasquid replied:
Yes that is definitely a possibility to investigate. There are pros and cons to doing it manually vs compile time. For example, doing it as a manual step gives you a way to control the flow of changes into the project and a git diff. Where, doing it at compile time gives the advantage of not having to worry about a manual step. I'm definitely open to changing this aspect of it, especially if the manual `generate-code` step becomes a burden. I added it to the https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs as feature to further investigate. Thanks for the feedback :)

gigasquid replied:
That's a good idea. I added the detail of this to the Clojure Package Contribution Needs - thanks :)

nswamy replied:
@gigasquid I am also concerned that we are merging generated code, we do not want to merge massive amounts of generated code -- this is also a moving target as and when new operators are added or modified. The proposal to only have the interface/documentation also helps us to have a config driven implementation. Do you think you can prioritize this before 1.3?

gigasquid replied:
@nswamy Of course. I can work on that next.

nswamy replied:
thanks so much :), you rock!

yzhliu commented on contrib/clojure-package/src/org/apache/clojure_mxnet/gen/symbol.clj:
```diff
@@ -0,0 +1,10940 @@
+
+
+
+
```
same question as that in ndarray.

gigasquid replied:
It is the same process as described in the ndarray.

gigasquid commented:
Thanks @yzhliu for your review. If you have any other questions, please let me know 😄

szha commented:
@gigasquid and team, congrats! It's been a long way since @gigasquid pitched the idea of having a clojure frontend for mxnet since NIPS last year, and I'm thrilled to see this becoming the reality now. This effort opens up mxnet to the clojure enthusiasts, and I hope that we will attract more people from this community and see more interesting work coming soon :)

larroy commented:
Let's merge this in unless somebody has additional comments.

gigasquid commented on contrib/clojure-package/project.clj:
```diff
@@ -0,0 +1,54 @@
+                 [t6/from-scala "0.3.0"]
+
+                 ;; Jars from Nexus
+                 ;[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.2.1"]
```
The CI build is taking it from master when the jars are installed locally with `make scalainstall`.  I kept the maven jars lines in there as well commented out so that it is easy for a user to switch back and forth.
